learn/parse_http.go

```diff
@@ -145,14 +144,12 @@ func ParseHTTP(elem akinet.ParsedNetworkContent) (*PartialWitness, error) {
 		}
 
 		if err != nil {
-			// Just log an error instead of returning an error so users can see the
-			// other parts of the endpoint in the spec rather than an empty spec.
-			// https://app.clubhouse.io/akita-software/story/1898/juan-s-payload-problem
-			telemetry.RateLimitError("unparsable body", err)
-			printer.Debugf("skipping unparsable body: %v\n", err)
-		} else if bodyData != nil {
-			datas = append(datas, bodyData)
+			// When the body is unparsable even after attempting fallback decompressions,
+			// we will try to capture the body as a string and indicate parsing error in the body meta
+			bodyData = captureUnparsableBody(rawBody, contentType, statusCode)
```

Comment:
I think this does not operate correctly when (a) the body is compressed, but (b) it is invalid JSON, because you go back to `rawBody`.  Is there any way to test this out?  (For example, you could run apidump locally and dump to file, if that's still supported, and look at the witness after setting up a call like this.)

Probably we need to handle decompression in captureUnparsableBody, or else set up a new `decodeStream` and use that instead.

Reply:
Thanks for catching this case! 
I have handled it by creating two readers out of the decoded stream. Do you think this is inefficient? I was not able to think of a better solution 🤔 
I have also added a test case for this issue and verified it with and without the fix.

Reply:
It looks to me like decodeBody returns a reader, and we can call it twice if we need to without any downside.  All of the compression types use a streaming model, as does the character-set transform.  The underlying MemView can support multiple readers without any problem.

(I don't recall in what circumstances we get the body already decompressed?)

However, the only way we get to this point is by calling attemptDecompress(), which _does_ attempt to make a full copy.  I think that when it fails it will usually not read very many bytes (unless the problem is on-the-wire data corruption.)  So probably we should not try to re-use that buffer as our primary means.

If I were to re-do this entirely, I would set up each as a separate stream to avoid the "attempt to read the whole thing before parsing" problem in the attempted uncompression attempts.

````
rawBody -> uncompress -> decode -> parseBody
rawBody -> attempted uncompression #1 -> decode -> parseBody
rawBody -> attempted uncompression #2 -> decode -> parseBody
...
rawBody -> uncompress -> decode -> captureUnparsableBody
```

It is not clear whether we should be decoding in the last line (our new pipeline) or not.  If we do not, then we might get invalid Unicode in our string (so we should be using Bytes if we are not already?  Did that come up in design review?)

But with that as our guide, it seems like the right thing here is to just start over with rawBody.CreateReader(), which is lightweight, and decodeBody, which is streaming, for the input to captureUnparseableBody.   I do not think it is worth saving the attempted uncompression outcome, even if successful, but if it is then we should remember which algorithm worked in our first pass and set up a new stream, rather than create another buffered copy as your current code does.

Please do this instead of creating the extra copy.  It is OK if we don't handle the combination of (compression method was not specified) and (body did not parse after attempted compression seemed to work), I'm primarily interested in catching (compression method was valid) and (body was invalid JSON) rather than other less-likely iterations.

The current code does not do a good job of verifying whether compression was actually used in the first pipeline or not.  If it does, I don't think we should do the expensive retry at all.  But, with each chain of readers it seems a bit hard to figure out which step blew up without a bunch of extra work.  Maybe the best solution is *not* to attempt to uncompression trials if there appears to be a valid compression header.

Reply:
There is room for a lot of improvements in this flow. 
For now, I have addressed these concerns by creating a copy from the memview and passing it through the decode body again. This should work for now? 

```
bodyStream := rawBody.CreateReader()
decodeStream, _ := decodeBody(headers, bodyStream, bodyDecompressed)
bodyData = captureUnparsableBody(decodeStream, contentType, statusCode)
```

Reply:
I can create a JIRA for follow-up improvements. Works?

Reply:
JIRA for the same https://postmanlabs.atlassian.net/browse/POA-2575